oc new-app throws segfault on invalid build strategy

`oc` crashes when trying to instantiate a template with an invalid buildconfig strategy:

```
$ cat tmp.yaml
apiVersion: v1
kind: Template
objects:
- apiVersion: v1
  kind: BuildConfig
  spec:
    strategy:
      type: foobar
$ oc new-app --file tmp.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x375d82e]

goroutine 1 [running]:
github.com/openshift/origin/pkg/generate/app/cmd.(*AppConfig).buildTemplates(0xc42072b880, 0xc4202b87a0, 0x1, 0x1, 0xc420590a50, 0xc420590990, 0xc4205909f0, 0x7109b20, 0xc420fb22e0, 0x0, ...)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/generate/app/cmd/newapp.go:510 +0x4fe
github.com/openshift/origin/pkg/generate/app/cmd.(*AppConfig).Run(0xc42072b880, 0xc4212138a0, 0xc420db7e30, 0xc42118d8e0)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/generate/app/cmd/newapp.go:833 +0xc10
github.com/openshift/origin/pkg/oc/cli/cmd.(*NewAppOptions).RunNewApp(0xc420150320, 0x0, 0x2)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/oc/cli/cmd/newapp.go:305 +0x33c
github.com/openshift/origin/pkg/oc/cli/cmd.NewCmdNewApplication.func1(0xc420498b40, 0xc420fb3d20, 0x0, 0x2)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/oc/cli/cmd/newapp.go:226 +0x10a
github.com/openshift/origin/vendor/github.com/spf13/cobra.(*Command).execute(0xc420498b40, 0xc420fb3c00, 0x2, 0x2, 0xc420498b40, 0xc420fb3c00)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/github.com/spf13/cobra/command.go:603 +0x22b
github.com/openshift/origin/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420260fc0, 0xc42008e010, 0xc420260fc0, 0x4)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/github.com/spf13/cobra/command.go:689 +0x339
github.com/openshift/origin/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420260fc0, 0x2, 0xc420260fc0)
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
        /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/cmd/oc/oc.go:36 +0x1ae
```

##### Version

```
$ oc version
oc v3.7.1+ab0f056
kubernetes v1.7.6+a08f5eeb62
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://127.0.0.1:8443
openshift v3.7.1+ab0f056
kubernetes v1.7.6+a08f5eeb62
```

##### Current Result

Segfaults.

##### Expected Result

 Doesn't segfault, gives clean error.

--- a/pkg/build/util/util.go
+++ b/pkg/build/util/util.go
@@ -375,6 +375,8 @@ func SetBuildConfigEnv(buildConfig *buildapi.BuildConfig, env []kapi.EnvVar) {
 		oldEnv = &buildConfig.Spec.Strategy.CustomStrategy.Env
 	case buildConfig.Spec.Strategy.JenkinsPipelineStrategy != nil:
 		oldEnv = &buildConfig.Spec.Strategy.JenkinsPipelineStrategy.Env
+	default:
+		return
 	}
 	*oldEnv = env
 }
@@ -392,6 +394,8 @@ func SetBuildEnv(build *buildapi.Build, env []kapi.EnvVar) {
 		oldEnv = &build.Spec.Strategy.CustomStrategy.Env
 	case build.Spec.Strategy.JenkinsPipelineStrategy != nil:
 		oldEnv = &build.Spec.Strategy.JenkinsPipelineStrategy.Env
+	default:
+		return
 	}
 	*oldEnv = env
 }

--- a/test/cmd/newapp.sh
+++ b/test/cmd/newapp.sh
@@ -48,6 +48,9 @@ os::cmd::expect_success 'oc new-app https://github.com/sclorg/ruby-ex'
 os::cmd::expect_success_and_not_text 'oc describe bc/ruby-ex' 'BuildConfigInstantiateFailed'
 os::cmd::expect_success 'oc delete all -l app=ruby-ex'
 
+# Ensure that an invalid build strategy in a template does not throw a segmentation fault
+os::cmd::expect_success_and_not_text 'oc new-app --file test/testdata/invalid-build-strategy.yaml --dry-run' 'invalid memory address or nil pointer dereference'
+
 # test that imagestream references across imagestreams do not cause an error
 os::cmd::try_until_success 'oc get imagestreamtags ruby:2.3'
 os::cmd::expect_success 'oc create -f test/testdata/newapp/imagestream-ref.yaml'

--- a/test/testdata/invalid-build-strategy.yaml
+++ b/test/testdata/invalid-build-strategy.yaml
@@ -1,0 +1,8 @@
+apiVersion: v1
+kind: Template
+objects:
+- apiVersion: v1
+  kind: BuildConfig
+  spec:
+    strategy:
+      type: foobar

